core/utils/historical-data/src/linear.rs

```diff
@@ -0,0 +1,417 @@
+	fn remove(&mut self, index: usize) {
+		self.remove_range(index, index + 1);
+	}
+
+	fn remove_range(&mut self, index: usize, end: usize) {
+		if end == 0 {
+			return;
+		}
+		let len = self.len();
+		if len <= end - index && index == 0 {
+			self.clear();
+			return;
+		}
+		// eager removal is costy, running some gc impl
+		// can be interesting.
+		let elt_start = self.index_element(index);
+		let start_ix = self.index_start();
+		let elt_end = if end == len {
+			start_ix
+		} else {
+			self.index_element(end) 
+		};
+		let delete_size = elt_end - elt_start;
+		for _ in elt_start..elt_end {
+			let _ = self.0.to_mut().remove(elt_start);
+		}
+		let start_ix = start_ix - delete_size;
+
+		let len = len - (end - index);
+		for i in index..end {
+			let pos = i + (end - index);
+			if pos < len {
+				let old_value = self.read_le_usize(start_ix + pos * SIZE_BYTE_LEN);
+				self.write_le_usize(start_ix + i * SIZE_BYTE_LEN, old_value - delete_size);
+			}
+		}
+		let end_index = start_ix + len * SIZE_BYTE_LEN;
+		self.write_le_usize(end_index - SIZE_BYTE_LEN, len);
+		self.0.to_mut().truncate(end_index);
+
+	}
+
+	pub(crate) fn get_state(&self, index: usize) -> HistoricalValue<&[u8], u64> {
+		let start_ix = self.index_element(index);
+		let len = self.len();
+		let end_ix = if index == len - 1 {
+			self.index_start()
+		} else {
+			self.index_element(index + 1)
+		};
+		let state = self.read_le_u64(start_ix);
+		HistoricalValue {
+			value: &self.0[start_ix + SIZE_BYTE_LEN..end_ix],
+			index: state,
+		}
+	}
+
+}
+
+const EMPTY_SERIALIZED: [u8; SIZE_BYTE_LEN] = [0u8; SIZE_BYTE_LEN];
+const DEFAULT_VERSION: u8 = 1;
+const DEFAULT_VERSION_EMPTY_SERIALIZED: [u8; SIZE_BYTE_LEN + 1] = {
+	let mut buf = [0u8; SIZE_BYTE_LEN + 1];
+	buf[0] = DEFAULT_VERSION;
+	buf
+};
+
+impl<'a, F: SerializedConfig> Default for Serialized<'a, F> {
+	fn default() -> Self {
+		Serialized(Cow::Borrowed(F::empty()), PhantomData)
+	}
+}
+
+impl<'a, F> Into<Serialized<'a, F>> for &'a[u8] {
+	fn into(self) -> Serialized<'a, F> {
+		Serialized(Cow::Borrowed(self), PhantomData)
+	}
+}
+
+impl<F> Into<Serialized<'static, F>> for Vec<u8> {
+	fn into(self) -> Serialized<'static, F> {
+		Serialized(Cow::Owned(self), PhantomData)
+	}
+}
+
+// Utility function for basis implementation.
+impl<'a, F: SerializedConfig> Serialized<'a, F> {
+	
+	// Index at end, also contains the encoded size
+	fn index_start(&self) -> usize {
+		let nb_ix = self.len();
+		if nb_ix == 0 { return F::version_len(); }
+		let end = self.0.len();
+		end - (nb_ix * SIZE_BYTE_LEN)
+	}
+
+	fn index_element(&self, position: usize) -> usize {
+		if position == 0 {
+			return F::version_len();
+		}
+		let i = self.index_start() + (position - 1) * SIZE_BYTE_LEN;
+		self.read_le_usize(i)
+	}
+
+	fn slice_copy(&mut self, start_from: usize, start_to: usize, size: usize) {
+		self.0.to_mut().copy_within(start_from..start_from + size, start_to);
+	}
+
+	// Usize encoded as le u64 (for historical value).
```

Comment:
```suggestion
	// Read u64 encoded as le (for history index).
```

what do you mean by Usize encoded as le u64, history index is u64 itself